### ISSUE TITLE
Unsupported intrinsic functions

### ISSUE THREAD

**Issue by hoffa** [aws/serverless-application-model#2533]
Creating this to help track and manage intrinsics-related issues.

### The problem

Some properties have limited support for [intrinsic functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) (e.g. `Fn::If`, `Fn::FindInMap`, etc.). Using unsupported intrinsic functions in such properties causes the deployment to fail.

### Why it's a problem

AWS SAM is a [AWS CloudFormation macro](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-macros.html); it receives a [SAM template](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-template-anatomy.html) as input (as-is, along with the intrinsic functions), and returns a [CloudFormation template](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-anatomy.html) which is then deployed by CloudFormation.

This means that unless the intrinsic function is statically resolvable (e.g. it only refers to [CloudFormation parameters](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html)), SAM has no way of resolving the intrinsic function. It's not an issue for properties that are passed as-is to properties of the underlying CloudFormation resources, but it becomes an issue when SAM must know the value for its transform logic (see e.g. the [`Enabled` property](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-schedule.html#sam-function-schedule-enabled) of `Schedule` events).

**Comment by tom-mcclintock** [comment]
Chiming in to say I came here via #2377 and am very interested to see this get supported.

In my use case, I use `!Ref` to set VPC endpoint IDs:
```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Parameters:
  VPCE:
    Type: String

Resources:
  ApiGateway:
    Type: AWS::Serverless::Api
    Properties:
      StageName: somestagename
      EndpointConfiguration:
        Type: PRIVATE
        VPCEndpointIds:
          - !Ref VPCE
```

Running `sam validate` yields an error, but also ends with the green message `... is a valid SAM Template`:
```bash
Template schema validation reported the following errors: [Resources.ApiGateway.Properties.EndpointConfiguration.VPCEndpointIds.0] {'Ref': 'VPCE'} is not of type 'string'
template.yaml is a valid SAM Template
```

So, it is not a blocker, but it would be nice to eliminate this error :). Thanks for working on this.

**Comment by berzi** [comment]
Is anyone working on this? I'm seemingly unable to define API endpoints in nested stacks because importing the API (AWS::Serverless::Api) from the parent stack doesn't work (resulting in "RestApiId must be a valid reference to an 'AWS::Serv
erless::Api' resource in same template" and I suspect this is the reason. It seems like a rather vital feature to have.

**Comment by hoffa** [comment]
> Is anyone working on this? I'm seemingly unable to define API endpoints in nested stacks because importing the API (AWS::Serverless::Api) from the parent stack doesn't work (resulting in "RestApiId must be a valid reference to an 'AWS::Serv erless::Api' resource in same template" and I suspect this is the reason. It seems like a rather vital feature to have.

At a glance unsure whether it's related; if you create a new issue with reproducible steps we can assist you there. As for intrinsic function resolution as described in this issue, there are no current plans to address this in the SAM transform, as `AWS::LanguageExtensions` already does it.

**Comment by tariromukute** [comment]
I am coming from issue #1159. I am failing to get the deployment to work for LayerVersion. Please see a snap of the template.yaml below.

```yaml
utilitiesLayer:
    Type: AWS::Serverless::LayerVersion
    Properties:
      LayerName: utilities-layer
      Description: Utility layer
      ContentUri: !Ref UtilitiesPath
      CompatibleRuntimes:
        - nodejs12.x
      LicenseInfo: "MIT"
      RetentionPolicy: Retain
```

The error I am getting is as below due to `ContentUri: !Ref UtilitiesPath`

```
"FAILED" Status: FAILED. Reason: Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [utilitiesLayerfc27b2746a] is invalid. 'ContentUri' is not a valid S3 Uri of the form 's3://bucket/key' with optional versionId query parameter.
```

Any ideas on how I can resolve this, or any work arounds?

**Comment by mgorski-mg** [comment]
> I am coming from issue #1159. I am failing to get the deployment to work for LayerVersion. Please see a snap of the template.yaml below.
> 
> ```yaml
> utilitiesLayer:
>     Type: AWS::Serverless::LayerVersion
>     Properties:
>       LayerName: utilities-layer
>       Description: Utility layer
>       ContentUri: !Ref UtilitiesPath
>       CompatibleRuntimes:
>         - nodejs12.x
>       LicenseInfo: "MIT"
>       RetentionPolicy: Retain
> ```
> 
> The error I am getting is as below due to `ContentUri: !Ref UtilitiesPath`
> 
> ```
> "FAILED" Status: FAILED. Reason: Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [utilitiesLayerfc27b2746a] is invalid. 'ContentUri' is not a valid S3 Uri of the form 's3://bucket/key' with optional versionId query parameter.
> ```
> 
> Any ideas on how I can resolve this, or any work arounds?

If UtilitiesPath is S3Bucket then !Ref is resolved to it's ID(name). More info [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html#aws-properties-s3-bucket-return-values).
The ContentUri field requires Uri: s3://.
What you probably need is
```
!Sub s3://${UtilitiesPath}
 ```

**Comment by tariromukute** [comment]
Thanks @mgorski-mg `UtilitiesPath` is a path to the folder with the layers. My Parameters sections looks like below.

```
AWSEndpoint:
    Type: String
    Description: The endpoint to use for AWS client. Useful when we want request to go to localstack
    Default: None
  UtilitiesPath:
    Type: String
    Description: The base path to utilities. Needed for running inside docker with docker compose. This should reference the base path inside docker container
    Default: "utilities-layer/"
```

If I put `ContentUri: utilities-layer/` it works fine. However I need to parametrise it because when I try to run this with docker compose the path needs to reference the working directory. You can see the discussion on this here https://github.com/aws/aws-sam-cli/issues/2837#issuecomment-1430149431

**Comment by mgorski-mg** [comment]
@tariromukute oh I see 🙈 that's strange them. Then it's probably connected to this issue, your right. 
You can go with this `AWS: LanguageExtension`. I checked it some time ago and it is doing the job.

**Comment by tariromukute** [comment]
Hey @mgorski-mg for responding quickly. I tried adding `AWS::LanguageExtensions` the error above is with `AWS::LanguageExtensions` and like below.

```
AWSTemplateFormatVersion: "2010-09-09"
Transform:
  - AWS::LanguageExtensions
  - AWS::Serverless-2016-10-31
```

When I remove `AWS::LanguageExtensions` I get a different but similar error, see the error below

```txt
Error: Failed to create changeset for the stack: rvw-stitch-tr-app, ex: Waiter ChangeSetCreateComplete failed: Waiter encountered a terminal failure state: For expression "Status" we matched expected path: "FAILED" Status: FAILED. Reason: Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [utilitiesLayer3b82bbbef7] is invalid. 'ContentUri' requires Bucket and Key properties to be specified.
```

**Comment by mgorski-mg** [comment]
@tariromukute i know i know now. I edited my previous comment.

**Comment by tariromukute** [comment]
> @tariromukute i know i know now. I edited my previous comment.

Seeing the edit now, thanks @mgorski-mg. Seems like it resolves the issue for some properties.

**Comment by hoffa** [comment]
@tariromukute The issue seems to be that the intrinsic function must be resolved before deployment (local paths are unsupported by CloudFormation; under the hood SAM CLI uploads the assets and creates an intermediary template with the final S3 URIs), and (1) the SAM transform does not attempt to resolve the intrinsic function locally, and (2) the SAM CLI does not support running `AWS::LanguageExtensions` locally. The limitation is briefly mentioned in https://github.com/aws/serverless-application-model/issues/2533#issue-1405166577.

**Comment by tariromukute** [comment]
Thank you @hoffa for the explanation. I went through the comment, however I wasn't able to understand what the workaround for this issue is. How can I use `aws cloudformation` directly to deploy a project that I initialised with `sam init`?

**Comment by hoffa** [comment]
@tariromukute You have to use SAM CLI to deploy any templates with local paths (such as most `sam init` templates), as SAM CLI needs to upload the assets first.

As for intrinsic functions that should resolve to local paths, unless the property implements some ad-hoc intrinsic function resolution, there's no great workaround apart from hardcoding. This is something we could explore in SAM CLI so that e.g. it also resolves intrinsic functions before passing to the SAM transform.

**Comment by tariromukute** [comment]
Hi @hoffa, thank you for clarifying the above. Is there an issue or work being done on this (we could explore in SAM CLI so that e.g. it also resolves intrinsic functions before passing to the SAM transform) or there is need for this to be created. This will make development and deployment easier. We tend to have javascript developers that do not have a lot of cloud experience and knowledge. Allowing SAM applications to be set up end-to-end locally using docker-compose (configure SAM app and local stack etc) will enable the developers to work on the lambda code without needing to know or deal with the cloud and SAM configurations.

**Comment by hoffa** [comment]
@tariromukute If I'm understanding correctly the problem you're encountering is `!Ref` to local paths in `AWS::Serverless::LayerVersion` not working. I was able to reproduce it, and confirmed it works with `CodeUri` in `AWS::Lambda::Function`. I've created https://github.com/aws/aws-sam-cli/issues/4767 for the issue. There's not much else that can be done on the SAM transform side as the intrinsic resolution to the local path must happen before the SAM transform receives it.

**Comment by tariromukute** [comment]
> @tariromukute If I'm understanding correctly the problem you're encountering is `!Ref` to local paths in `AWS::Serverless::LayerVersion` not working. I was able to reproduce it, and confirmed it works with `CodeUri` in `AWS::Lambda::Function`. I've created [aws/aws-sam-cli#4767](https://github.com/aws/aws-sam-cli/issues/4767) for the issue. There's not much else that can be done on the SAM transform side as the intrinsic resolution to the local path must happen before the SAM transform receives it. While not ideal, as workaround in the meantime you could use `sed` or similar to replace the text.

Awesome. Thanks @hoffa for reproducing the issue and for creating an issue for it. I will try the `sed` route on the CI pipeline.

**Comment by smaly-amazon** [comment]
If the AWS::LanguageExtensions transform is the solution, is that transform available in all regions where the serverless transform is available?

**Comment by davidchatz** [comment]
I've been using the language extension to workaround this issue with using `!If` in `AutoPublishAlias`, but this does not work with the new `sam list` command, which outputs: Error: [InvalidResourceException('SnsTopicConsumer', "'AutoPublishAlias' must be a string or a Ref to a template parameter")] ('SnsTopicConsumer', "'AutoPublishAlias' must be a string or a Ref to a template parameter")

**Comment by ozcsdog** [comment]
I am coming from https://github.com/aws/serverless-application-model/issues/1616.
Is there any way to set string to whole Step Functions TimeoutSeconds? Just like TimeoutSecondsPath for Task.
Now the TimeoutSeconds is only for int, and can't get string from path.

**Comment by anashakt** [comment]
I just now ran into same situation while trying to pass "TimeoutSeconds" as integer in `definitionSubstitutions`

**Comment by carlosleon-wl** [comment]
Hi, I'm having a very simple problem:
```yaml
      Events:
        Status:
          Type: HttpApi
          Properties:
            ApiId: !Ref HttpApi
            Method: get
            Path: !Sub "/${Version}/status"
```
I cannot use `!Sub` to substitute `Version` parameter.
This is a long-standing problem according to what I've seen. Is there anything I can do apart from hardcoding the `Version`?

**Comment by moelasmar** [comment]
@carlosleon-wl .. you can use the workaround mentioned in this issue.

### Add the `AWS::LanguageExtensions` transform

The [`AWS::LanguageExtensions` transform](https://aws.amazon.com/blogs/mt/introducing-new-language-extensions-in-aws-cloudformation/) will resolve intrinsic functions if the value is known when [`Transform`s](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-macros.html) are run.

Replace:

```yaml
Transform: AWS::Serverless-2016-10-31
```

With:

```yaml
Transform:
  - AWS::LanguageExtensions
  - AWS::Serverless-2016-10-31
```

**Comment by Uday599** [comment]
`Failed to create the changeset: Waiter ChangeSetCreateComplete failed: Waiter encountered a terminal failure state: For expression "Status" we matched expected path: "FAILED" Status: FAILED. Reason: SSM Secure reference is not supported in: [AWS::ApiGateway::ApiKey/Properties/Value]`

```
  PrivateApiGatewayApiKey:
    Type: AWS::ApiGateway::ApiKey
    Condition: ActiveEnvironment
    DependsOn:
      - PrivateApiGatewayApiliveStage
    Properties:
      Description: API Key for Metrics API V2 ECS
      Enabled: true
      Name:
        Fn::If:
          - Master
          - !Sub ${ServiceName}-${EnvironmentName}
          - !Sub ${ServiceName}-${BranchName}
      Value: !Sub "{{resolve:ssm-secure:/metrics/${EnvironmentName}/api-key:1}}"
      StageKeys:
        - RestApiId: !Ref PrivateApiGatewayApi
          StageName: !Ref aliasName
```



Any idea how do we resolve this error?
